### PR TITLE
Feature/add and check signatures consensus

### PR DESCRIPTION
**Pull request by n-hutton**

### CODE REVIEW

File: libs/core/include/core/fetch_backward.hpp [hunk 1]
```diff
@@ -50,3 +50,18 @@
 #if defined(__GNUC__)
 #pragma GCC diagnostic pop
 #endif
+
```
Comment (n-hutton) [inline libs/core/include/core/fetch_backward.hpp]: Unrelated but might be useful

File: apps/constellation/constellation.cpp [hunk 2]
```diff
@@ -121,6 +121,9 @@ uint16_t LookupLocalPort(Manifest const &manifest, ServiceIdentifier::Type servi
 std::shared_ptr<ledger::DAGInterface> GenerateDAG(std::string const &db_name, bool load_on_start,
                                                   Constellation::CertificatePtr certificate)
 {
+  FETCH_UNUSED(db_name);
```
Comment (ejfitzgerald) [inline apps/constellation/constellation.cpp]: Why is this necessary?

File: libs/beacon/include/beacon/block_entropy.hpp [hunk 3]
```diff
@@ -28,27 +28,35 @@ namespace beacon {
...
+  // The hash of the above (when new committee) note, this could be implicit. Is not
+  // serialized.
+  Digest digest;
+
+  // In the case of a new cabinet, personal signatures of the hash
+  // from qual members
+  Confirmations confirmations;
+
+  // Signature of the previous entropy, used as the entropy
+  GroupSignature group_signature = GroupSignature{};
```
Comment (ejfitzgerald) [inline libs/beacon/include/beacon/block_entropy.hpp]:
```suggestion
  GroupSignature group_signature{};
```

Comment (ejfitzgerald) [reply libs/beacon/include/beacon/block_entropy.hpp]: If needed at all?

File: libs/beacon/src/beacon_manager.cpp [hunk 4]
```diff
@@ -644,14 +644,15 @@ bool BeaconManager::Verify(Signature const &signature)
...
  */
-bool BeaconManager::Verify(std::string const &group_public_key, MessagePayload const &message,
-                           std::string const &signature)
+bool BeaconManager::Verify(byte_array::ConstByteArray const &group_public_key,
+                           MessagePayload const &            message,
+                           byte_array::ConstByteArray const &signature)
 {
   PublicKey tmp;
-  tmp.setStr(group_public_key);
+  tmp.setStr(std::string(group_public_key));
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_manager.cpp:657]:
```suggestion
  tmp.setStr(static_cast<std::string>(group_public_key));
```

Comment (ejfitzgerald) [reply libs/beacon/src/beacon_manager.cpp:657]: (I would prefer)

File: libs/beacon/src/beacon_manager.cpp [hunk 5]
```diff
@@ -644,14 +644,15 @@ bool BeaconManager::Verify(Signature const &signature)
...
 
   Signature tmp2;
-  tmp2.setStr(signature);
+  tmp2.setStr(std::string(signature));
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_manager.cpp:660]:
```suggestion
  tmp2.setStr(static_cast<std::string?(signature));
```

File: libs/beacon/src/beacon_service.cpp [hunk 6]
```diff
@@ -340,8 +340,28 @@ BeaconService::State BeaconService::OnCompleteState()
   beacon_entropy_last_generated_->set(index);
   beacon_entropy_generated_total_->add(1);
 
-  // Populate the block entropy structure appropriately TODO(HUT): this.
-  // block_entropy_being_created_->group_signature = active_exe_unit_->manager.GroupSignature();
+  // Populate the block entropy structure appropriately
+  block_entropy_being_created_->group_signature =
+      active_exe_unit_->manager.GroupSignature().getStr();
+
+  std::cerr << "pre-verify 0" << std::endl;  // DELETEME_NH
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_service.cpp]: Delete?

File: libs/core/include/core/fetch_backward.hpp (hunk 1 above)
Comment (ejfitzgerald) [reply libs/core/include/core/fetch_backward.hpp]: Worth adding to logging instead?

File: libs/crypto/tests/unit/mcl_dkg_tests.cpp [hunk 7]
```diff
@@ -29,7 +29,7 @@ using namespace fetch::byte_array;
 
 TEST(MclTests, BaseMcl)
 {
-  fetch::crypto::mcl::details::MCLInitialiser();
+  // fetch::crypto::mcl::details::MCLInitialiser();
```
Comment (ejfitzgerald) [inline libs/crypto/tests/unit/mcl_dkg_tests.cpp]: Remove these?

File: libs/ledger/src/chain/block_coordinator.cpp [hunk 8]
```diff
@@ -1106,20 +1112,25 @@ BlockCoordinator::State BlockCoordinator::OnReset()
 
   if (next_block_)
   {
+    std::cerr << "wasn " << next_block_->body.block_number << std::endl;  // DELETEME_NH
```
Comment (ejfitzgerald) [inline libs/ledger/src/chain/block_coordinator.cpp]: Remove?

File: libs/ledger/src/chain/block_coordinator.cpp [hunk 9]
```diff
@@ -1106,20 +1112,25 @@ BlockCoordinator::State BlockCoordinator::OnReset()
...
+    std::cerr << "wasc" << std::endl;  // DELETEME_NH
     block = current_block_.get();
   }
   else
   {
     FETCH_LOG_ERROR(LOGGING_NAME, "Unable to find a previously executed block!");
   }
 
+  // TODO(HUT): figure out where/when to do this
+  // block->UpdateDigest();
```
Comment (ejfitzgerald) [inline libs/ledger/src/chain/block_coordinator.cpp]: Already done after collecting final merkle tree hash?

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Seems good to me just a couple of niggles

File: libs/beacon/include/beacon/block_entropy.hpp (hunk 3 above)
Comment (n-hutton) [reply libs/beacon/include/beacon/block_entropy.hpp]: A relic from when I tried to use the MCL stuff direct (a mistake). Good spot will change.

File: libs/beacon/src/beacon_manager.cpp (hunk 5 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_manager.cpp:660]: this is constructing a string, not casting it

File: libs/beacon/src/beacon_manager.cpp (hunk 4 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_manager.cpp:657]: see below

File: libs/core/include/core/fetch_backward.hpp (hunk 1 above)
Comment (n-hutton) [reply libs/core/include/core/fetch_backward.hpp]: ok

File: libs/crypto/src/mcl_dkg.cpp [hunk 10]
```diff
@@ -33,6 +33,8 @@ namespace mcl {
 
 std::atomic<bool> details::MCLInitialiser::was_initialised{false};
 
+details::MCLInitialiser execute_me_before_runtime;
```
Comment (n-hutton) [inline libs/crypto/src/mcl_dkg.cpp]: Note: so long as this is linked, it should call the MCL init just before main is called. This might cause undefined behaviour if any MCL functions are called in other constructors so let's avoid this.

File: libs/ledger/src/chain/block_coordinator.cpp (hunk 8 above)
Comment (n-hutton) [reply libs/ledger/src/chain/block_coordinator.cpp]: Uhhh yes. I would say good spot but it was like a sore thumb

File: libs/ledger/src/chain/block_coordinator.cpp (hunk 9 above)
Comment (n-hutton) [reply libs/ledger/src/chain/block_coordinator.cpp]: There is a corner case when exceptions are thrown, this might not be true. Will open ticket.

File: libs/ledger/src/consensus/consensus.cpp [hunk 11]
```diff
@@ -398,6 +430,37 @@ NextBlockPtr Consensus::GenerateNextBlock()
...
+
+  return fetch::crypto::Verifier::Verify(block.body.miner_id, block.body.hash,
+                                         block.miner_signature);
+}
+
+/**
+ * Given a block entropy, determine whether it has been signed off on
+ * by enough qualified stakers.
+ */
+bool Consensus::EnoughQualSigned(BlockEntropy const &block_entropy) const
```
Comment (n-hutton) [inline libs/ledger/src/consensus/consensus.cpp:453]: Holding off on this part until your stake changes are in @ejfitzgerald

File: apps/constellation/constellation.cpp (hunk 2 above)
Comment (n-hutton) [reply apps/constellation/constellation.cpp]: It's not, I think I deleted it at the same time you commented ahah

File: libs/beacon/src/beacon_service.cpp (hunk 6 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_service.cpp]: done

File: libs/beacon/src/dkg_output.cpp [hunk 12]
```diff
@@ -22,7 +22,6 @@ using fetch::beacon::DkgOutput;
 
 DkgOutput::DkgOutput()
 {
-  fetch::crypto::mcl::details::MCLInitialiser();
```
Comment (n-hutton) [inline libs/beacon/src/dkg_output.cpp]: Since we call this in constellation main it is probably safe not to include these @jinmannwong

**Review by ejfitzgerald — DISMISSED** [review]

File: libs/beacon/src/beacon_manager.cpp (hunk 5 above)
Comment (ejfitzgerald) [reply libs/beacon/src/beacon_manager.cpp:660]: So does the above, but I am not too phased either way

**Review by ejfitzgerald — APPROVED** [review]
